Small scale anaerobic digestion system

osuphobia opened the pull request:
#### Summary
Features "Small scale anaerobic digestion system"

#### Purpose of change

Added an anaerobic digestion furniture, organic_waste requirement, biogas, biogas tank and cooker, and relevant recipes.
Borrowed and modified fermenting vat functions. Get biogas and fermented liquid from anaerobic digester.
Close https://github.com/CleverRaven/Cataclysm-DDA/pull/72828

#### Describe the solution

Spreading manure on crops recycles the nutrients, but as it decomposes it releases methane. And lots of it. Agriculture is the [largest source](https://www.epa.gov/ghgemissions/overview-greenhouse-gases#methane) of methane emissions globally. Each year, methane from livestock manure has the [warming equivalent](https://www.sciencedirect.com/science/article/pii/S0301479712006603) of 240 million tons of carbon dioxide, or the same as a year of emissions from 52 million cars.

One of the most effective ways of mitigating this powerful greenhouse gas is to capture it and burn it as natural gas. The Chinese government has been incentivising farmers to do this for decades, [so are the US government](https://www.epa.gov/agstar/what-epa-doing-agstar).

For now, I only managed to introduce a [small scale anaerobic digestion system](https://sswm.info/arctic-wash/module-4-technology/further-resources-wastewater-treatment/anaerobic-digestion-%28small-scale%29) to dda.

Dig a deep pit and build an anaerobic digester on it, mix dungs and water, pour them into the digester.
After one month, the biogas generating process will start, you can get biogas from the digester every 24 hours.
After two months, you can get fermented fertilizer liquid, and make liquid fertilizer from it with a simple step.

I think it's more realistic than the plan in https://github.com/CleverRaven/Cataclysm-DDA/pull/72828, as well as the exsisting simple recipe( so I removed it, the complex one is left though).

![China_biogas_digester_training-768x576](https://github.com/CleverRaven/Cataclysm-DDA/assets/78858975/adbdddd6-bc53-4c6f-bba5-609f874e6540)
Training programme in biogas digester construction in Shandong province (Image: Karen Mancl)
![SASSE 1998 and TUTORVISTA 2010  Three types of fixed-dome biogas reactor  Each biogas plant](https://github.com/CleverRaven/Cataclysm-DDA/assets/78858975/4c132376-0362-4423-ac4d-4be9f143f22a)

#### Describe alternatives you've considered

Add compost heap item? But it may attract a lot of scavengers and monsters in this apocalypse world.

#### Testing

Compiled and tested locally.

<img width="1187" alt="ssads" src="https://github.com/CleverRaven/Cataclysm-DDA/assets/78858975/652f2526-bd83-41b6-bc74-1a52d23e890d">

<img width="510" alt="123" src="https://github.com/CleverRaven/Cataclysm-DDA/assets/78858975/81cc8232-86c6-45a7-b2e1-e8ecb1b14dfd">

#### Additional context

Add industrial anaerobic digester in the future for farms and faction camp. AgSTAR in dda!

osuphobia committed:
Add anaerobic digester, organic_waste requirement,
biogas, biogas tank and cooker, and relevant recipes.

osuphobia committed:
Borrowed and modified fermenting vat functions.
Get biogas and fermented liquid from anaerobic digester.

osuphobia committed:
Modified the density of dungs.
Add water_sewage to organic_waste.

osuphobia committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Hyperseeker commented:
This is definitely an interesting approach. I'd like to see where it goes.

One thing that doesn't quite jive, at least intuitively, is only requiring 60 bricks and a single metal sheet to produce a thousand-liter-capable container. Even the pit in the first image seems to be using far more than 60 bricks alone. Plus, having it only take 3 hours seems like superhuman speed.

Have these values been vetted in any way, or are they a placeholder?

osuphobia commented:
> Have these values been vetted in any way, or are they a placeholder?

@Hyperseeker 

You are right, I just forgot to change them to resonable values, considering changing the amount of bricks to 600 and time needed to 2 days. For the single metal sheet, I was imaging a metal cover when starting to write this PR, but now I just want to remove it from the recipe, as the metal tank servers as the gasholder sits on top of the digester.

Besides that, I shouldn't modify the density of dungs in this PR, should have spilt the modification to another single PR.

And there are also some issues needed to be fixed as well, so I'll mark this as draft.

osuphobia committed:
Revert the modification of the density of dungs.

osuphobia committed:
Add new words biogas/biogases
to tools/spell_checker/dictionary.txt.

osuphobia committed:
Change the recipe of the digester to make it more
resonable.

Karol1223 commented:
So I get it I can close #65413 then?

osuphobia commented:
> So I get it I can close #65413 then?

@Karol1223 
The function in my PR works as intended, but there still are some weird things mainly in JSON, minor but quite annoying. 
Had a glance, codes in your PR need to be rewritten, because of a recent code migration https://github.com/CleverRaven/Cataclysm-DDA/pull/72954.
Since that your PR is not finished yet, yes you can wait my one to be finished and then close yours.

osuphobia committed:
Fix biogas tank.

osuphobia committed:
Clang-tidy

Rocket-F-1024 commented:
Is 600 bricks a bit too much? A whole pallet of bricks in the game is only 400. And actually 1000 liters (1 cubic meter) isn't really that big, it's actually only 1 meter on each side if it's a standard cube.
The other thing is that the structure presumably doesn't need to be heat resistant, so perhaps it could be built using adobe bricks?

osuphobia committed:
Add crafting_pseudo_item fake_digester_tank to
anaerobic digester, and use it in the recipes of
fermentable_liquid_mixture.
Add CONTAIN level 1 to the requirement of simple
liquid fertilizer recipe.

osuphobia commented:
> Is 600 bricks a bit too much? A whole pallet of bricks in the game is only 400. And actually 1000 liters (1 cubic meter) isn't really that big, it's actually only 1 meter on each side if it's a standard cube. The other thing is that the structure presumably doesn't need to be heat resistant, so perhaps it could be built using adobe bricks?

@Rocket-F-1024 
An anaerobic digester like the one in the photo above do need that much. Maybe we need less in the game, but I'm not familiar with the game-to-reality conversion ratio.
And please note that the volume of the anaerobic digester is larger than 1000 liter.
For the adobe brick, since that anaerobic digester needs to be (almost) airtight, it may not be a good idea IMO.

osuphobia committed:
Clang-tidy

Qrox commented on src/iexamine.cpp:
```diff
@@ -3712,6 +3715,281 @@ void iexamine::fvat_full( Character &you, const tripoint &examp )
+                    const int count = compost_i.count();
+                    const time_point gas_birthday = compost_i.birthday();
+                    if( !max_gas_gatherable ) {
+                        add_msg( _( "No biogass gathered." ) );
```
```suggestion
                        add_msg( _( "No biogas gathered." ) );
```

osuphobia committed:
Update src/iexamine.cpp

Co-authored-by: Jianxiang Wang (王健翔) <qrox@sina.com>

ADekema commented:
1.Why can you only put use butchery refuse and feces in the digestor? You could pretty much use any kind of organic material as long as it rots (even whole bodies). You should probably include all types of food (good way to recycle rotten food), by products (from milling, oil pressing, cheesemaking, etc) and butchering products. The only exeptions should be woody materials, bone and chitin. You could techically digest those things it would simply take much longer so wouldn´t be worth it. 

2. Would it be possible to just make propane tanks be able to hold biogas instead of ¨Crafting¨ them into biogas tanks? Als seems strange that gas tanks can only hold a certain type of gas. You would think that a gas tank would be capable of holding any type of gas. Might be some exemptions and even if there aren´t doing a gas tank overhaul would be a differant PR,

3. Can you really use biogas simply as is? Isn´t there a lot of moistere, carbondioxide, and nasty smelling adolterants mixed into the gas? Isn´t there like a whole series of steps that need to be undertaken to upgrade the gas to burnable/cooking/primitive generator grade and even more work to upgrade it to methane/vehicle/gas generator grade?

osuphobia commented:
> 1. Why can you only put use butchery refuse and feces in the digestor? You could pretty much use any kind of organic material as long as it rots (even whole bodies). You should probably include all types of food (good way to recycle rotten food), by products (from milling, oil pressing, cheesemaking, etc) and butchering products. The only exeptions should be woody materials, bone and chitin. You could techically digest those things it would simply take much longer so wouldn´t be worth it.
> 2. Would it be possible to just make propane tanks be able to hold biogas instead of ¨Crafting¨ them into biogas tanks? Als seems strange that gas tanks can only hold a certain type of gas. You would think that a gas tank would be capable of holding any type of gas. Might be some exemptions and even if there aren´t doing a gas tank overhaul would be a differant PR,
> 3. Can you really use biogas simply as is? Isn´t there a lot of moistere, carbondioxide, and nasty smelling adolterants mixed into the gas? Isn´t there like a whole series of steps that need to be undertaken to upgrade the gas to burnable/cooking/primitive generator grade and even more work to upgrade it to methane/vehicle/gas generator grade?

1. Of course, kitchen and garden wastes are suitable materials, while woody materials are not, as they contain a lot of lignin which resists anaerobic fermentation and should therefore not be used unless pre-composted and preferably chopped. For the food stuff and corpse, I can deal with them afterwards, but the digestion functions need to be somewhat rewritten then. Now I'm just sitting here praying for getting this PR merged.
2. The pocket type needs to be container to gather biogas from digester, otherwise the digestion functions need to be totally rewritten (and I don't know how to do it btw).
3. Yes and yes. Biogas does contain impurities (mainly carbon dioxide), you can purify it to get methane, but you can also simply use it as it is. https://www.youtube.com/watch?v=c_Jtl-VhkKI Here's a video of a DIY biogas digester made of a plastic barrel, biogas is used by the gas stove as it is. Purification of biogas would be a differant PR.
